File: module3/class1/example1-m3p1e1-intensive-io-demonstration.cs
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Diagnostics;
using System.Collections.Generic;

namespace IntensiveIODemonstration
{
    // ====================================================================
    // CONFIGURATION VARIABLES - EASY TO MODIFY FOR DIFFERENT SCENARIOS
    // ====================================================================
    public static class IOConfig
    {
        public const int MIN_FILE_SIZE_KB = 100;        // Minimum file size in KB
        public const int MAX_FILE_SIZE_KB = 500;        // Maximum file size in KB
        public const int WRITE_CHUNK_SIZE = 4096;       // Write chunk size in bytes (4KB)
        public const int READ_BUFFER_SIZE = 1024;       // Read buffer size in bytes (1KB)
        public const int READ_REPETITIONS = 3;          // How many times to read each file
        public const int STATISTICS_INTERVAL = 10;      // Show statistics every N cycles
        public const int CYCLE_DELAY_MS = 500;          // Delay between cycles in milliseconds
        public const int WRITE_DELAY_MS = 1;            // Delay between write operations in ms
        public const int READ_DELAY_MS = 1;             // Delay between read operations in ms
        public const int GC_DEMO_INTERVAL = 5;          // Perform GC demo every N cycles
        public const int GC_WASTE_OBJECTS = 10;         // Number of waste objects for GC demo
        public const int GC_WASTE_SIZE_MB = 1;          // Size of each waste object in MB
        public const string BASE_FILENAME = "intensive_io_file_";  // Base name for temp files
    }
    // ====================================================================

    class IntensiveIODemo
    {
        private readonly string baseFileName;
        private int fileCounter;
        private long totalBytesWritten;
        private long totalBytesRead;
        private int totalOperations;
        private readonly Stopwatch stopwatch;
        private readonly Random random;

        public IntensiveIODemo()
        {
            baseFileName = IOConfig.BASE_FILENAME;
            fileCounter = 0;
            totalBytesWritten = 0;
            totalBytesRead = 0;
            totalOperations = 0;
            stopwatch = Stopwatch.StartNew();
            random = new Random();
        }

        private string GenerateLargeContent(int sizeKB)
        {
            var sb = new StringBuilder();
            
            // Header information
            sb.AppendLine("=== INTENSIVE I/O DEMONSTRATION DATA ===");
            sb.AppendLine($"File #{fileCounter} - Timestamp: {DateTimeOffset.Now.ToUnixTimeMilliseconds()}");
            sb.AppendLine($"Size: {sizeKB} KB");
            sb.AppendLine(new string('=', 50));
            sb.AppendLine();

            // Calculate remaining size needed
            int currentSize = Encoding.UTF8.GetByteCount(sb.ToString());
            int targetSize = sizeKB * 1024;
            int remainingSize = targetSize - currentSize;

            // Fill with random data
            var chars = new char[remainingSize];
            for (int i = 0; i < remainingSize; i++)
            {
                if (i % 80 == 79)
                {
                    chars[i] = '\n'; // Add line breaks for readability
                }
                else
                {
                    chars[i] = (char)random.Next(65, 91); // A-Z characters
                }
            }

            sb.Append(chars);
            return sb.ToString();
        }

        private void PerformIntensiveWrite()
        {
            string fileName = $"{baseFileName}{fileCounter}.tmp";
            
            // Generate large content (using configured size range)
            int fileSize = random.Next(IOConfig.MIN_FILE_SIZE_KB, IOConfig.MAX_FILE_SIZE_KB + 1);
            string content = GenerateLargeContent(fileSize);
            
            // Perform multiple write operations to the same file
            using (var fileStream = new FileStream(fileName, FileMode.Create, FileAccess.Write, FileShare.None))
            using (var writer = new BinaryWriter(fileStream))
            {
                byte[] contentBytes = Encoding.UTF8.GetBytes(content);
                
                // Write in chunks to create more I/O operations
                for (int i = 0; i < contentBytes.Length; i += IOConfig.WRITE_CHUNK_SIZE)
                {
                    int currentChunkSize = Math.Min(IOConfig.WRITE_CHUNK_SIZE, contentBytes.Length - i);
                    byte[] chunk = new byte[currentChunkSize];
                    Array.Copy(contentBytes, i, chunk, 0, currentChunkSize);
                    
                    writer.Write(chunk);
                    fileStream.Flush(true); // Force immediate write to disk
                    
                    totalBytesWritten += currentChunkSize;
                    totalOperations++;
                    
                    // Small delay to make operations visible
                    Thread.Sleep(IOConfig.WRITE_DELAY_MS);
                }
            }
            
            Console.WriteLine($"WRITE: Created {fileName} ({fileSize} KB)");
        }

        private void PerformIntensiveRead()
        {
            string fileName = $"{baseFileName}{fileCounter}.tmp";
            
            // Read the file in small chunks multiple times
            for (int readAttempt = 0; readAttempt < IOConfig.READ_REPETITIONS; readAttempt++)
            {
                if (File.Exists(fileName))
                {
                    using (var fileStream = new FileStream(fileName, FileMode.Open, FileAccess.Read, FileShare.Read))
                    using (var reader = new BinaryReader(fileStream))
                    {
                        long fileSize = fileStream.Length;
                        
                        // Read in small chunks
                        byte[] buffer = new byte[IOConfig.READ_BUFFER_SIZE];
                        int bytesRead;
                        
                        while ((bytesRead = reader.Read(buffer, 0, buffer.Length)) > 0)
                        {
                            totalBytesRead += bytesRead;
                            totalOperations++;
                            
                            // Small delay to make operations visible
                            Thread.Sleep(IOConfig.READ_DELAY_MS);
                        }
                        
                        Console.WriteLine($"READ #{readAttempt + 1}: {fileName} ({fileSize} bytes)");
                    }
                }
            }
        }

        private void DeleteTemporaryFile()
        {
            string fileName = $"{baseFileName}{fileCounter}.tmp";
            
            try
            {
                if (File.Exists(fileName))
                {
                    File.Delete(fileName);
                    Console.WriteLine($"DELETE: Removed {fileName}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error deleting {fileName}: {ex.Message}");
            }
        }

        private void DisplayStatistics()
        {
            TimeSpan elapsed = stopwatch.Elapsed;
            
            Console.WriteLine();
            Console.WriteLine(new string('=', 60));
            Console.WriteLine("INTENSIVE I/O STATISTICS");
            Console.WriteLine(new string('=', 60));
            Console.WriteLine($"Running time: {elapsed.TotalSeconds:F1} seconds");
            Console.WriteLine($"Total operations: {totalOperations:N0}");
            Console.WriteLine($"Files processed: {fileCounter:N0}");
            Console.WriteLine($"Total bytes written: {totalBytesWritten / 1024.0 / 1024.0:F2} MB");
            Console.WriteLine($"Total bytes read: {totalBytesRead / 1024.0 / 1024.0:F2} MB");
            Console.WriteLine($"Operations per second: {(elapsed.TotalSeconds > 0 ? totalOperations / elapsed.TotalSeconds : 0):F2}");
            Console.WriteLine($"Memory usage: {GC.GetTotalMemory(false) / 1024.0 / 1024.0:F2} MB");
            Console.WriteLine(new string('=', 60));
        }

        private void PerformGarbageCollectionDemo()
        {
            // Intentionally create and abandon large objects to demonstrate GC pressure
            var wastefulList = new List<byte[]>();
            
            for (int i = 0; i < IOConfig.GC_WASTE_OBJECTS; i++)
            {
                // Create large byte arrays that will be collected
                byte[] wasteData = new byte[IOConfig.GC_WASTE_SIZE_MB * 1024 * 1024]; // Configurable MB each
                wastefulList.Add(wasteData);
            }
            
            // Clear the list to make objects eligible for GC
            wastefulList.Clear();
            
            // Force garbage collection to demonstrate memory pressure
            GC.Collect();
            GC.WaitForPendingFinalizers();
            GC.Collect();
            
            Console.WriteLine("GC: Performed aggressive garbage collection");
        }

        public void Run()
        {
            Console.WriteLine("=== INTENSIVE I/O DEMONSTRATION (C#) ===");
            Console.WriteLine("This program will perform excessive disk I/O operations");
            Console.WriteLine("WARNING: This will stress your disk subsystem!");
            Console.WriteLine("Press any key to stop the demonstration...");
            Console.WriteLine(new string('-', 50));

            bool continueRunning = true;
            
            // Start a background task to monitor for key presses
            var keyMonitorTask = System.Threading.Tasks.Task.Run(() =>
            {
                Console.ReadKey(true);
                continueRunning = false;
                Console.WriteLine("\nStopping demonstration...");
            });

            while (continueRunning)
            {
                fileCounter++;
                
                Console.WriteLine($"\n--- Cycle #{fileCounter} ---");
                
                try
                {
                    // Perform intensive I/O operations
                    PerformIntensiveWrite();
                    PerformIntensiveRead();
                    DeleteTemporaryFile();
                    
                    // Occasionally demonstrate memory pressure
                    if (fileCounter % IOConfig.GC_DEMO_INTERVAL == 0)
                    {
                        PerformGarbageCollectionDemo();
                    }
                    
                    // Display statistics every configured interval
                    if (fileCounter % IOConfig.STATISTICS_INTERVAL == 0)
                    {
                        DisplayStatistics();
                    }
                    
                    // Brief pause between cycles
                    Thread.Sleep(IOConfig.CYCLE_DELAY_MS);
                }
                catch (Exception ex)
                {
                    Console.WriteLine($"Error in cycle {fileCounter}: {ex.Message}");
                    Thread.Sleep(1000); // Longer pause on error
                }
            }
            
            // Clean up any remaining temporary files
            CleanupTemporaryFiles();
            
            // Final statistics
            DisplayStatistics();
            
            Console.WriteLine("\nDemonstration completed. Press any key to exit...");
            Console.ReadKey(true);
        }

        private void CleanupTemporaryFiles()
        {
            try
            {
                string[] tempFiles = Directory.GetFiles(".", $"{IOConfig.BASE_FILENAME}*.tmp");
                foreach (string file in tempFiles)
                {
                    File.Delete(file);
                    Console.WriteLine($"CLEANUP: Removed {file}");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error during cleanup: {ex.Message}");
            }
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                var demo = new IntensiveIODemo();
                demo.Run();
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Fatal error: {ex.Message}");
                Console.WriteLine("Press any key to exit...");
                Console.ReadKey(true);
            }
        }
    }
}

File: module3/class1/example1-m3p1e1-intensive-io-demonstration.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <chrono>
#include <thread>
#include <iomanip>
#include <random>
#include <sstream>
#include <conio.h>  // For _kbhit() on Windows

// ====================================================================
// CONFIGURATION VARIABLES - EASY TO MODIFY FOR DIFFERENT SCENARIOS
// ====================================================================
const size_t MIN_FILE_SIZE_KB = 100;        // Minimum file size in KB
const size_t MAX_FILE_SIZE_KB = 500;        // Maximum file size in KB
const size_t WRITE_CHUNK_SIZE = 4096;       // Write chunk size in bytes (4KB)
const size_t READ_BUFFER_SIZE = 1024;       // Read buffer size in bytes (1KB)
const int READ_REPETITIONS = 3;             // How many times to read each file
const int STATISTICS_INTERVAL = 10;         // Show statistics every N cycles
const int CYCLE_DELAY_MS = 500;             // Delay between cycles in milliseconds
const int WRITE_DELAY_MICROSECONDS = 100;   // Delay between write operations
const int READ_DELAY_MICROSECONDS = 50;     // Delay between read operations
const std::string BASE_FILENAME = "intensive_io_file_";  // Base name for temp files
// ====================================================================

class IntensiveIODemonstration {
private:
    std::string baseFileName;
    int fileCounter;
    size_t totalBytesWritten;
    size_t totalBytesRead;
    int totalOperations;
    std::chrono::high_resolution_clock::time_point startTime;
    
    // Generate large dummy data
    std::string generateLargeContent(size_t sizeKB) {
        std::stringstream ss;
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> dis(65, 90); // A-Z characters
        
        size_t totalChars = sizeKB * 1024;
        ss << "=== INTENSIVE I/O DEMONSTRATION DATA ===\n";
        ss << "File #" << fileCounter << " - Timestamp: " 
           << std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::high_resolution_clock::now().time_since_epoch()).count() 
           << "\n";
        ss << "Size: " << sizeKB << " KB\n";
        ss << std::string(50, '=') << "\n\n";
        
        // Fill with random data to reach desired size
        for (size_t i = ss.str().length(); i < totalChars; ++i) {
            if (i % 80 == 79) {
                ss << '\n';  // Add line breaks for readability
            } else {
                ss << static_cast<char>(dis(gen));
            }
        }
        
        return ss.str();
    }
    
    void performIntensiveWrite() {
        std::string fileName = baseFileName + std::to_string(fileCounter) + ".tmp";
        
        // Generate large content (using configured size range)
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_int_distribution<> sizeDis(MIN_FILE_SIZE_KB, MAX_FILE_SIZE_KB);
        size_t fileSize = sizeDis(gen);
        
        std::string content = generateLargeContent(fileSize);
        
        // Perform multiple write operations to the same file
        std::ofstream file(fileName, std::ios::binary | std::ios::trunc);
        if (file.is_open()) {
            // Write in chunks to create more I/O operations
            for (size_t i = 0; i < content.length(); i += WRITE_CHUNK_SIZE) {
                size_t currentChunkSize = std::min(WRITE_CHUNK_SIZE, content.length() - i);
                file.write(content.c_str() + i, currentChunkSize);
                file.flush(); // Force immediate write to disk
                totalBytesWritten += currentChunkSize;
                totalOperations++;
                
                // Small delay to make operations visible
                std::this_thread::sleep_for(std::chrono::microseconds(WRITE_DELAY_MICROSECONDS));
            }
            file.close();
        }
        
        std::cout << "WRITE: Created " << fileName << " (" << fileSize << " KB)" << std::endl;
    }
    
    void performIntensiveRead() {
        std::string fileName = baseFileName + std::to_string(fileCounter) + ".tmp";
        
        // Read the file in small chunks multiple times
        for (int readAttempt = 0; readAttempt < READ_REPETITIONS; ++readAttempt) {
            std::ifstream file(fileName, std::ios::binary);
            if (file.is_open()) {
                file.seekg(0, std::ios::end);
                size_t fileSize = file.tellg();
                file.seekg(0, std::ios::beg);
                
                // Read in small chunks
                char buffer[READ_BUFFER_SIZE];
                while (file.read(buffer, sizeof(buffer)) || file.gcount() > 0) {
                    totalBytesRead += file.gcount();
                    totalOperations++;
                    
                    // Small delay to make operations visible
                    std::this_thread::sleep_for(std::chrono::microseconds(READ_DELAY_MICROSECONDS));
                }
                file.close();
                
                std::cout << "READ #" << (readAttempt + 1) << ": " << fileName 
                         << " (" << fileSize << " bytes)" << std::endl;
            }
        }
    }
    
    void deleteTemporaryFile() {
        std::string fileName = baseFileName + std::to_string(fileCounter) + ".tmp";
        if (std::remove(fileName.c_str()) == 0) {
            std::cout << "DELETE: Removed " << fileName << std::endl;
        }
    }
    
    void displayStatistics() {
        auto currentTime = std::chrono::high_resolution_clock::now();
        auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(currentTime - startTime);
        
        std::cout << "\n" << std::string(60, '=') << std::endl;
        std::cout << "INTENSIVE I/O STATISTICS" << std::endl;
        std::cout << std::string(60, '=') << std::endl;
        std::cout << "Running time: " << elapsed.count() << " seconds" << std::endl;
        std::cout << "Total operations: " << totalOperations << std::endl;
        std::cout << "Files processed: " << fileCounter << std::endl;
        std::cout << "Total bytes written: " << std::fixed << std::setprecision(2) 
                 << (totalBytesWritten / 1024.0 / 1024.0) << " MB" << std::endl;
        std::cout << "Total bytes read: " << std::fixed << std::setprecision(2) 
                 << (totalBytesRead / 1024.0 / 1024.0) << " MB" << std::endl;
        std::cout << "Operations per second: " << std::fixed << std::setprecision(2) 
                 << (elapsed.count() > 0 ? totalOperations / static_cast<double>(elapsed.count()) : 0) << std::endl;
        std::cout << std::string(60, '=') << std::endl;
    }
    
public:
    IntensiveIODemonstration() : 
        baseFileName(BASE_FILENAME), 
        fileCounter(0), 
        totalBytesWritten(0), 
        totalBytesRead(0), 
        totalOperations(0) {
        startTime = std::chrono::high_resolution_clock::now();
    }
    
    void run() {
        std::cout << "=== INTENSIVE I/O DEMONSTRATION (C++) ===" << std::endl;
        std::cout << "This program will perform excessive disk I/O operations" << std::endl;
        std::cout << "WARNING: This will stress your disk subsystem!" << std::endl;
        std::cout << "Press any key to stop the demonstration..." << std::endl;
        std::cout << std::string(50, '-') << std::endl;
        
        while (true) {
            // Check if user wants to stop
            if (_kbhit()) {
                char ch = _getch();
                std::cout << "\nStopping demonstration..." << std::endl;
                break;
            }
            
            fileCounter++;
            
            std::cout << "\n--- Cycle #" << fileCounter << " ---" << std::endl;
            
            // Perform intensive I/O operations
            performIntensiveWrite();
            performIntensiveRead();
            deleteTemporaryFile();
            
            // Display statistics every configured interval
            if (fileCounter % STATISTICS_INTERVAL == 0) {
                displayStatistics();
            }
            
            // Brief pause between cycles
            std::this_thread::sleep_for(std::chrono::milliseconds(CYCLE_DELAY_MS));
        }
        
        // Final statistics
        displayStatistics();
        
        std::cout << "\nDemonstration completed. Press any key to exit..." << std::endl;
        _getch();
    }
};

int main() {
    try {
        IntensiveIODemonstration demo;
        demo.run();
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        std::cout << "Press any key to exit..." << std::endl;
        _getch();
        return 1;
    }
    
    return 0;
}
